libsignal-protocol/src/crypto/mod.rs

```diff
@@ -0,0 +1,381 @@
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError>;
+}
+
+/// A simple vtable ([`signal_crypto_provider`]) and set of trampolines to let C
+/// use our [`Crypto`] trait object.
+pub(crate) struct CryptoProvider {
+    pub(crate) vtable: signal_crypto_provider,
+    state: Pin<Box<State>>,
+}
+
+impl CryptoProvider {
+    pub fn new<C: Crypto + 'static>(crypto: C) -> CryptoProvider {
+        // we need a double-pointer because C doesn't do fat pointers
+        let mut state: Pin<Box<State>> = Box::pin(State(Box::new(crypto)));
+
+        let vtable = signal_crypto_provider {
+            user_data: state.as_mut().get_mut() as *mut State as *mut c_void,
+            random_func: Some(random_func),
+            hmac_sha256_init_func: Some(hmac_sha256_init_func),
+            hmac_sha256_update_func: Some(hmac_sha256_update_func),
+            hmac_sha256_final_func: Some(hmac_sha256_final_func),
+            hmac_sha256_cleanup_func: Some(hmac_sha256_cleanup_func),
+            sha512_digest_init_func: Some(sha512_digest_init_func),
+            sha512_digest_update_func: Some(sha512_digest_update_func),
+            sha512_digest_final_func: Some(sha512_digest_final_func),
+            sha512_digest_cleanup_func: Some(sha512_digest_cleanup_func),
+            encrypt_func: Some(encrypt_func),
+            decrypt_func: Some(decrypt_func),
+        };
+
+        CryptoProvider { vtable, state }
+    }
+
+    pub fn state(&self) -> &dyn Crypto { &*self.state.0 }
+}
+
+struct State(Box<dyn Crypto>);
+
+struct HmacContext(RefCell<Box<dyn Sha256Hmac>>);
+
+struct DigestContext(RefCell<Box<dyn Sha512Digest>>);
+
+unsafe extern "C" fn random_func(
+    data: *mut u8,
+    len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!data.is_null());
+    assert!(!user_data.is_null());
+
+    let user_data = &*(user_data as *const State);
+    let buffer = slice::from_raw_parts_mut(data, len);
+    user_data.0.fill_random(buffer).into_code()
+}
+
+unsafe extern "C" fn hmac_sha256_cleanup_func(
+    hmac_context: *mut c_void,
+    _user_data: *mut c_void,
+) {
+    assert!(!hmac_context.is_null());
+
+    let hmac_context: Box<HmacContext> =
+        Box::from_raw(hmac_context as *mut HmacContext);
+    drop(hmac_context);
+}
+
+unsafe extern "C" fn hmac_sha256_final_func(
+    hmac_context: *mut c_void,
+    output: *mut *mut signal_buffer,
+    _user_data: *mut c_void,
+) -> i32 {
+    // just to make sure that the c ffi gave us a valid buffer to write to.
+    assert!(!output.is_null());
+    assert!(!hmac_context.is_null());
+
+    let hmac_context = &*(hmac_context as *const HmacContext);
+
+    match hmac_context.0.borrow_mut().finalize() {
+        Ok(hmac) => {
+            let buffer = Buffer::from(hmac);
+            *output = buffer.into_raw();
+            sys::SG_SUCCESS as c_int
+        },
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn hmac_sha256_init_func(
+    hmac_context: *mut *mut c_void,
```

Comment:
i always wonder the use of that arg `hmac_context` and how can we apply that to our implementation, so i think that is a good use of it, also it is now easy to see what is going on in the crypto module